[X86] lowerFPToIntToFP - handle UI2FP on AVX512VL targets and i64 types on AVX512DQ targets

KavinTheG opened the pull request:
fixes llvm#160111

KavinTheG committed:
[X86] Add optimizations for fp to signed & unsigned i64/32 conversions. (#160111)

Extends lowerFPToIntToFP to support i64/i32 conversions on AVX VLX + DQ
and DQ targets, as well as modifies appropriate test cases.

RKSimon reviewed (commented):
This seems more complicated than it should be - maybe just start with the AVX512DQ i64 handling in this patch?

RKSimon reviewed (commented):
Any idea why the unsigned tests in fp-int-fp-cvt.ll aren't affected on AVX512 targets?

RKSimon commented on llvm/test/CodeGen/X86/ftrunc.ll:
```diff
@@ -10,10 +10,8 @@ declare i64 @llvm.fptosi.sat.i64.f64(double)
-; SSE2-NEXT:    xorps %xmm0, %xmm0
-; SSE2-NEXT:    cvtsi2ss %rax, %xmm0
+; SSE2-NEXT:    cvttps2dq %xmm0, %xmm0
+; SSE2-NEXT:    cvtdq2ps %xmm0, %xmm0
```
suspicious?

KavinTheG replied:
I'll add an early out in the SSE2 handling if the int is unsigned.

KavinTheG commented:
@RKSimon, gentle ping for review.

KavinTheG commented on llvm/test/CodeGen/X86/fp-int-fp-cvt.ll:
```diff
@@ -143,11 +173,20 @@ define double @ucvtf64_i64(double %a0) {
-; AVX512-NEXT:    vcvttsd2usi %xmm0, %rax
-; AVX512-NEXT:    vcvtusi2sd %rax, %xmm15, %xmm0
-; AVX512-NEXT:    retq
+; AVX512-VL-LABEL: ucvtf64_i64:
```
My bad, I thought I was to focus on i64 handling for this patch. 

I can push a commit that removes the check for i64 like below to add optimization for `ucvtf64_i32`.
```
  if (Subtarget.hasVLX() && IntVT == MVT::i64) {
```
to 
```
  if (Subtarget.hasVLX()) {
```

KavinTheG commented:
I would still like to finish this PR so I would appreciate any review!

RKSimon reviewed (changes requested):
please can you fix the merge conflict?

KavinTheG committed:
Updated failing tests with python script

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -20186,7 +20186,9 @@ static SDValue lowerFPToIntToFP(SDValue CastToFP, const SDLoc &DL,
   // TODO: Allow FP_TO_UINT.
```
remove todo?

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -20198,22 +20200,60 @@ static SDValue lowerFPToIntToFP(SDValue CastToFP, const SDLoc &DL,
+  MVT VecSrcVT, VecIntVT, VecVT;
+  unsigned NumElts;
+  unsigned SrcElts, VTElts;
+  // Some conversions are only legal with uniform vector sizes on AVXDQ.
```
```suggestion
  // Some conversions are only legal with uniform vector sizes on AVX512DQ.
```

RKSimon commented on llvm/test/CodeGen/X86/fp-int-fp-cvt.ll:
```diff
@@ -16,11 +16,26 @@ define double @scvtf64_i32(double %a0) {
+; AVX512-NOVL-NEXT:    vcvtdq2pd %ymm0, %zmm0
+; AVX512-NOVL-NEXT:    # kill: def $xmm0 killed $xmm0 killed $zmm0
+; AVX512-NOVL-NEXT:    vzeroupper
+; AVX512-NOVL-NEXT:    retq
```
this is unnecessary - as only AVX instructions are required - we don't need AVX512 NO VLX special case for i32/f64 SINT_TO_FP

RKSimon reviewed (changes requested):
drop isint.ll changes

KavinTheG committed:
Dropped isint.ll changes and modified lowerFPToIntToFP to use AVX
instructions for i32/f64 SINT_TO_FP.

KavinTheG commented:
- dropped test cases added to isint.ll
- Modified `lowerFPToIntToFp` to use AVX isntructions for i32/f64 SINT_TO_FP.

RKSimon reviewed (changes requested):
Once #176919 lands, please can you merge against trunk to ensure avx512f-only targets are still OK

KavinTheG committed:
Modified guard if statement to ensure AVX512F-only targets are supported and updated fp-int-fp-cvt.ll test.

KavinTheG commented:
- Rebased upstream/main to include #176919. 
- Included suggest change to ensure support for AVX512F-only target.
- Updated `fp-int-fp-cvt.ll` test.

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into X86lowerFPToIntToFP-UI2FP

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -20216,22 +20217,56 @@ static SDValue lowerFPToIntToFP(SDValue CastToFP, const SDLoc &DL,
-  MVT VecVT = MVT::getVectorVT(VT, 128 / VTSize);
-
-  // We need target-specific opcodes if this is v2f64 -> v4i32 -> v2f64.
+  bool IsUnsigned = CastToInt.getOpcode() == ISD::FP_TO_UINT;
```
@KavinTheG we're missing a check that CastToFP has matching signedness - I have a patch incoming for #217355
